File: src/dexcom.ts
/**
 * dexcom.ts — minimal Dexcom Share follower client (EU / shareous1).
 *
 * Same approach that already works in GlukosRun / Viktcoachen:
 *  - Dexcom Share (real-time), NOT the official 3h-delayed API.
 *  - Two-step auth (AuthenticatePublisherAccount -> LoginPublisherAccountById).
 *  - Session cached in memory; re-auth on empty/failed read.
 *
 * Credentials come from env (never hard-coded, never committed). See .env.example.
 */

const APPLICATION_ID = "d89443d2-327c-4a6f-89e5-496bbb0317db"; // standard Share follower app id

const REGIONS: Record<string, string> = {
  eu: "https://shareous1.dexcom.com/ShareWebServices/Services",
  us: "https://share2.dexcom.com/ShareWebServices/Services",
};

export type Reading = {
  mmol: number;
  mgdl: number;
  trend: string;
  /** epoch ms of the reading (WT) */
  ts: number;
};

export type DexcomConfig = {
  username: string;
  password: string;
  region?: "eu" | "us";
};

function toMmol(mgdl: number): number {
  return Math.round((mgdl / 18.0182) * 10) / 10;
}

// Dexcom sometimes returns a bare quoted GUID; JSON.parse handles it, but we
// read as text and strip quotes to be robust to content-type quirks.
async function postForString(url: string, body: unknown): Promise<string> {
  const r = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json" },
    body: JSON.stringify(body),
  });
  const txt = (await r.text()).trim();
  if (!r.ok) throw new Error(`Dexcom ${r.status}: ${txt.slice(0, 200)}`);
  return txt.replace(/^"|"$/g, "");
}

export class DexcomShare {
  private base: string;
  private accountId: string | null = null;
  private sessionId: string | null = null;

  constructor(private cfg: DexcomConfig) {
    this.base = REGIONS[cfg.region ?? "eu"];
  }

  private async login(): Promise<void> {
    this.accountId = await postForString(`${this.base}/General/AuthenticatePublisherAccount`, {
      accountName: this.cfg.username,
      password: this.cfg.password,
      applicationId: APPLICATION_ID,
    });
    this.sessionId = await postForString(`${this.base}/General/LoginPublisherAccountById`, {
      accountId: this.accountId,
      password: this.cfg.password,
      applicationId: APPLICATION_ID,
    });
  }

  private async readRaw(minutes: number, maxCount: number): Promise<any[]> {
    if (!this.sessionId) await this.login();
    const url =
      `${this.base}/Publisher/ReadPublisherLatestGlucoseValues` +
      `?sessionId=${this.sessionId}&minutes=${minutes}&maxCount=${maxCount}`;
    const r = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json", Accept: "application/json" },
    });
    const txt = await r.text();
    if (!r.ok) throw new Error(`Dexcom read ${r.status}: ${txt.slice(0, 200)}`);
    let arr: any[] = [];
    try {
      arr = JSON.parse(txt);
    } catch {
      arr = [];
    }
    return arr;
  }

  /** Fetch up to `hours` of readings (oldest first). Retries once with fresh session. */
  async fetchSeries(hours = 6): Promise<Reading[]> {
    const minutes = hours * 60;
    const maxCount = Math.ceil((hours * 60) / 5) + 2;
    let raw = await this.readRaw(minutes, maxCount);
    if (raw.length === 0) {
      // empty often means the session was invalidated by another follower app
      this.sessionId = null;
      await this.login();
      raw = await this.readRaw(minutes, maxCount);
    }
    const readings = raw.map(parseReading).filter((x): x is Reading => x !== null);
    readings.sort((a, b) => a.ts - b.ts); // oldest -> newest
    return readings;
  }
}

function parseReading(x: any): Reading | null {
  if (!x || typeof x.Value !== "number") return null;
  const mgdl = x.Value;
  const trend = typeof x.Trend === "string" ? x.Trend : String(x.Trend ?? "");
  const ts = parseWt(x.WT ?? x.ST ?? x.DT);
  return { mgdl, mmol: toMmol(mgdl), trend, ts };
}

// Dexcom timestamps look like "/Date(1699999999999+0000)/"
function parseWt(wt: unknown): number {
  if (typeof wt !== "string") return Date.now();
  const m = wt.match(/\/Date\((\d+)/);
  return m ? Number(m[1]) : Date.now();
}

/** Least-squares slope (mmol/min) over the last `windowMin` minutes. */
export function slopePerMin(readings: Reading[], windowMin = 20): number {
  if (readings.length < 2) return 0;
  const tEnd = readings[readings.length - 1].ts;
  const pts = readings.filter((r) => tEnd - r.ts <= windowMin * 60 * 1000);
  if (pts.length < 2) return 0;
  const t0 = pts[0].ts;
  const xs = pts.map((p) => (p.ts - t0) / 60000);
  const ys = pts.map((p) => p.mmol);
  const n = xs.length;
  const sx = xs.reduce((a, b) => a + b, 0);
  const sy = ys.reduce((a, b) => a + b, 0);
  const sxx = xs.reduce((a, b) => a + b * b, 0);
  const sxy = xs.reduce((a, b, i) => a + b * ys[i], 0);
  const denom = n * sxx - sx * sx;
  return denom === 0 ? 0 : (n * sxy - sx * sy) / denom;
}

File: android/app/src/main/java/se/pixelmug/glucoseface/Dexcom.kt
package se.pixelmug.glucoseface

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

/** A glucose reading. */
data class Reading(val mmol: Double, val ts: Long)

/**
 * Minimal Dexcom Share follower client (EU/US) — same approach as GlukosRun /
 * Viktcoachen: real-time Share, two-step auth, session cached in memory.
 */
class Dexcom(private val user: String, private val pass: String, region: String = "eu") {
    private val base = if (region == "us")
        "https://share2.dexcom.com/ShareWebServices/Services"
    else "https://shareous1.dexcom.com/ShareWebServices/Services"
    private val appId = "d89443d2-327c-4a6f-89e5-496bbb0317db"
    private val http = OkHttpClient()
    private val JSON = "application/json".toMediaType()
    private var session: String? = null

    private fun postText(url: String, body: JSONObject): String {
        val req = Request.Builder().url(url)
            .post(body.toString().toRequestBody(JSON))
            .header("Accept", "application/json").build()
        http.newCall(req).execute().use { r ->
            val t = (r.body?.string() ?: "").trim()
            if (!r.isSuccessful) throw RuntimeException("Dexcom ${r.code}: ${t.take(160)}")
            return t.trim('"')
        }
    }

    private fun login() {
        val acct = postText("$base/General/AuthenticatePublisherAccount",
            JSONObject().put("accountName", user).put("password", pass).put("applicationId", appId))
        session = postText("$base/General/LoginPublisherAccountById",
            JSONObject().put("accountId", acct).put("password", pass).put("applicationId", appId))
    }

    private fun readRaw(minutes: Int, maxCount: Int): JSONArray {
        if (session == null) login()
        val url = "$base/Publisher/ReadPublisherLatestGlucoseValues?sessionId=$session&minutes=$minutes&maxCount=$maxCount"
        val req = Request.Builder().url(url).post("".toRequestBody(JSON)).build()
        http.newCall(req).execute().use { r ->
            val t = r.body?.string() ?: "[]"
            if (!r.isSuccessful) throw RuntimeException("Dexcom read ${r.code}")
            return try { JSONArray(t) } catch (e: Exception) { JSONArray() }
        }
    }

    /** Last `hours` of readings, oldest first. Retries once with a fresh session. */
    fun fetchSeries(hours: Int = 3): List<Reading> {
        val minutes = hours * 60
        val maxCount = (minutes / 5) + 2
        var raw = readRaw(minutes, maxCount)
        if (raw.length() == 0) { session = null; login(); raw = readRaw(minutes, maxCount) }
        val out = ArrayList<Reading>()
        for (i in 0 until raw.length()) {
            val o = raw.optJSONObject(i) ?: continue
            val mgdl = o.optDouble("Value", Double.NaN)
            if (mgdl.isNaN()) continue
            out.add(Reading(Math.round(mgdl / 18.0182 * 10) / 10.0, parseWt(o.optString("WT"))))
        }
        out.sortBy { it.ts }
        return out
    }

    private fun parseWt(wt: String?): Long {
        if (wt == null) return System.currentTimeMillis()
        val m = Regex("/Date\\((\\d+)").find(wt) ?: return System.currentTimeMillis()
        return m.groupValues[1].toLong()
    }
}
